File: lib/services/rt_membership_service.dart
import 'package:supabase_flutter/supabase_flutter.dart';

import '../models/rt_membership_model.dart';

class RtMembershipService {
  final SupabaseClient _supabase = Supabase.instance.client;

  Future<List<RtMembershipModel>> listMyMemberships() async {
    try {
      final result = await _supabase.rpc('list_my_rt_memberships');
      if (result == null) return [];
      final list = result as List;
      return list
          .map((e) => RtMembershipModel.fromMap(Map<String, dynamic>.from(e as Map)))
          .toList();
    } catch (_) {
      return _fallbackSingleMembership();
    }
  }

  Future<List<RtMembershipModel>> _fallbackSingleMembership() async {
    final userId = _supabase.auth.currentUser?.id;
    if (userId == null) return [];
    final profile = await _supabase
        .from('profiles')
        .select('rt_id, role')
        .eq('id', userId)
        .maybeSingle();
    final rtId = profile?['rt_id'] as String?;
    if (rtId == null) return [];
    final rt = await _supabase.from('rt_units').select().eq('id', rtId).maybeSingle();
    if (rt == null) return [];
    return [
      RtMembershipModel(
        rtId: rtId,
        rtName: rt['name'] as String? ?? 'RT',
        rtNumber: rt['rt_number'] as String? ?? '',
        rwNumber: rt['rw_number'] as String?,
        address: rt['address'] as String?,
        role: profile?['role'] as String? ?? 'warga',
        joinedAt: DateTime.now(),
        isActive: true,
      ),
    ];
  }

  Future<void> switchActiveRt(String rtId) async {
    await _supabase.rpc('switch_active_rt', params: {'p_rt_id': rtId});
  }

  Future<Map<String, dynamic>> joinRtByCode(String code) async {
    final result = await _supabase.rpc('join_rt_by_code', params: {
      'p_invite_code': code.trim(),
    });
    return Map<String, dynamic>.from(result as Map);
  }
}

File: mobile/src/services/rtMembershipService.ts
// Port dari lib/services/rt_membership_service.dart
import { supabase } from '../config/supabase';
import { RtMembership, rtMembershipFromMap } from '../types/membership';

export const rtMembershipService = {
  async listMyMemberships(): Promise<RtMembership[]> {
    try {
      const { data, error } = await supabase.rpc('list_my_rt_memberships');
      if (error) throw error;
      if (!data) return [];
      return (data as any[]).map(rtMembershipFromMap);
    } catch {
      return fallbackSingleMembership();
    }
  },

  async switchActiveRt(rtId: string): Promise<void> {
    const { error } = await supabase.rpc('switch_active_rt', { p_rt_id: rtId });
    if (error) throw error;
  },

  async joinRtByCode(code: string): Promise<Record<string, any>> {
    const { data, error } = await supabase.rpc('join_rt_by_code', {
      p_invite_code: code.trim(),
    });
    if (error) throw error;
    return (data ?? {}) as Record<string, any>;
  },
};

async function fallbackSingleMembership(): Promise<RtMembership[]> {
  const { data: userData } = await supabase.auth.getUser();
  const userId = userData.user?.id;
  if (userId == null) return [];

  const { data: profile } = await supabase
    .from('profiles')
    .select('rt_id, role')
    .eq('id', userId)
    .maybeSingle();
  const rtId = profile?.rt_id as string | undefined;
  if (rtId == null) return [];

  const { data: rt } = await supabase.from('rt_units').select().eq('id', rtId).maybeSingle();
  if (rt == null) return [];

  return [
    {
      rtId,
      rtName: (rt.name as string) ?? 'RT',
      rtNumber: (rt.rt_number as string) ?? '',
      rwNumber: (rt.rw_number as string) ?? null,
      address: (rt.address as string) ?? null,
      role: (profile?.role as string) ?? 'warga',
      joinedAt: new Date(),
      isActive: true,
    },
  ];
}
